proposals/inline-classes.md

```diff
@@ -0,0 +1,163 @@
+# Inline classes
+
+* **Type**: Design proposal
+* **Author**: Mikhail Zarechenskiy
+* **Contributors**: Andrey Breslav, Denis Zharkov, Roman Elizarov, Stanislav Erokhin
+* **Status**: Under consideration
+* **Prototype**: Implemented in Kotlin 1.2.30
+
+Discussion of this proposal is held in [this issue](https://github.com/Kotlin/KEEP/issues/104).
+
+## Summary
+
+Currently, there is no performant way to create wrapper for a value of a corresponding type. The only way is to create a usual class, 
+but the use of such classes would require additional heap allocations, which can be critical for many use cases.    
+
+We propose to support identityless inline classes that would allow to introduce wrappers for a values without additional overhead related 
+to additional heap allocations.
+
+## Motivation / use cases
+
+Inline classes allow to create wrappers for a value of a certain type and such wrappers would be fully inlined. 
+This is similar to type aliases but inline classes are not assignment-compatible with the corresponding underlying types.
+
+Use cases:
+
+- Unsigned types
+```
+inline class UInt(private val value: Int) { ... }
+inline class UShort(private val value: Short) { ... }
+inline class UByte(private val value: Byte) { ... }
+inline class ULong(private val value: Long) { ... }
+```
+
+- Native types like `size_t` for Kotlin/Native
+- Inline enum classes 
+    - Int enum for [Android IntDef](https://developer.android.com/reference/android/support/annotation/IntDef.html)
```

Comment:
How does this suppose to work?
Of course, you can create one inline class for each int, but it's not clear how this help with enum case.
Some kind enums based on inline classes or sealed would be really helpful tho
Are inline enums part of this proposal?

Reply:
I believe inline enum classes deserve another proposal, but I'll put here a bit information about them